Sources/JWTProvider/PayloadAuthenticationMiddleware.swift

```
@@ -64,22 +64,24 @@ public final class PayloadAuthenticationMiddleware<U: PayloadAuthenticatable>: M
 
         let filteredSigners = try self.signers(for: jwt)
 
-        // Try to use all the signers until one matches
-        var verified = false
+        // Default Error
+        var error: Swift.Error? = JWTError.signatureVerificationFailed
 
+        // Try to use all the signers until one matches
         for signer in filteredSigners {
 
             do {
-                _ = try req.jwt(verifyUsing: signer, and: claims)
-                verified = true
+                _ = try req.jwt(verifyUsing: signer, and: self.claims)
+                error = nil
                 break
-            } catch {
+            } catch let jwtError {
+                error = jwtError
                 continue
             }
         }
 
-        guard verified else {
-            throw JWTProviderError.noVerifiedJWT
```

I think it would be better to have the `noVerifiedJWT` case hold the original JWT error (or perhaps an array of all the errors thrown).

---

The problem with the previous implementation is that `noVerifiedJWT` overrides the jwt validation error.

It could fail because the token is invalid (which should end up in a 401) , or because claims are not validated (403).

Also, can't throw an array of errors, you'll need an error, that can contains many errors. In that case, which one of them would define the status code?

---

> Also, can't throw an array of errors, you'll need an error, that can contains many errors. In that case, which one of them would define the status code?

That seems to indicate some bad logic here then. 

```swift
let filteredSigners = try self.signers(for: jwt)
```

Is there a case when that call _should_ return multiple signers? Perhaps we can throw an error if multiple signers are available for a single JWT. Then it's much clearer how to proceed with the actual verification errors.

---

this is the only case causing that method to return an array with more than one element:

```
guard let kid = jwt.keyIdentifier else {
            // The token doesn't include a kid, try to verify using all the signers
            return Array(self.signers.values)
        }
```

We could remove that logic, but then jwt tokens without `kid` won't be verified anymore, so it won't be backwards compatible. I am not sure we want to enforce people to use `kid`, although it's good practice.

---

The current released version doesn't support `kid` at all, right? So we don't need to worry about backward compatibility.

I'd say we should require a kid iff there are multiple signers. If there's just one signer, then we don't need a kid and we just use the only available signer. 

Throwing away errors as this would require will very likely cause bugs. Also validating against multiple signers with each request could potentially be a performance issue.

---

If we require `kid` to be present, then `self.signers(for: jwt)` can be renamed to `self.signer(for: jwt)`.

Rest of the code is highly simplified, we don't need to loop and we don't need to catch and rethrow the errors.